File: lib/rpgmv/Parsers.js
'use strict';

const fs = require('fs');

function map_Parser(jData){

	let data = jData;
	let extractedText = [];

	//jump to a event id
	const codeTypes = [401, 405, 102, 108, 105, 101, 0];

	//Event loop
	let evLength = data["events"].length;
	for (let ev = 0; ev < evLength; ++ev){
		
		if (data["events"][ev] == null){continue;}
		let tmpArr = []
		
		//Page loop
		let paLength = data["events"][ev]["pages"].length;
		for (let pa = 0; pa < paLength; ++pa){
			
			// relatedChunk is an int to make it easier to check for repeats
			let relatedChunk = 0;
			let arrInArr = [];
			
			//List loop
			let liLength = data["events"][ev]["pages"][pa]["list"].length;
			for (let li = 0; li < liLength; ++li){
				
				let cType = data["events"][ev]["pages"][pa]["list"][li]["code"];
				let Param = data["events"][ev]["pages"][pa]["list"][li]["parameters"][0];
				let inden = data["events"][ev]["pages"][pa]["list"][li]["indent"];
				
				// Checks if cType is in codeTypes array and pushes it to an array
				if (codeTypes.indexOf(cType) > -1){
					
					if (cType == 108){
						if (Param.includes("text_indicator") == true){
							tmpArr.push([{"codetype":cType, "indent":inden, "parameters":Param}]);
							continue;
						}
						continue;
					}
					
					if (cType == 105 || cType == 101){
						relatedChunk += 1;
					}
					if (cType == 0 && arrInArr.length > 0){
						relatedChunk = 0;
						tmpArr.push(arrInArr)
						arrInArr = [];
						continue;
					}
					if (cType == 101 && relatedChunk > 1 || cType == 105 && relatedChunk > 1){
						relatedChunk = 1;
						tmpArr.push(arrInArr);
						arrInArr = [];
						continue;
					}
					
					if (cType == 102 && relatedChunk > 1 || cType == 102 && relatedChunk == 0){
						tmpArr.push([{"codetype":cType, "indent":inden, "parameters":Param}])
						}
					if (relatedChunk == 1 && cType != 101 && cType != 105 ){
						arrInArr.push({"codetype":cType, "indent":inden, "parameters":Param});
						continue;
					}

				}
			}	
		}
		extractedText.push({"id":ev, "list":tmpArr});
	}
	return extractedText;
}

function ce_Parser(jData){
	
	let data = jData;
	let extractedText = [];

	//jump to a event id
	const codeTypes = [401, 405, 102, 108, 105, 101, 0];
	
	//Event loop
	let evLength = data.length
	for (let ev = 0; ev < evLength; ++ev){
		if (data[ev] == null){continue;}
		let tmpArr = [];
		let arrInArr = [];
		let relatedChunk = 0
		//List loop
		let liLength = data[ev]["list"].length;
		for (let li = 0; li < liLength; ++li){
			
			let cType = data[ev]["list"][li]["code"];
			let Param = data[ev]["list"][li]["parameters"][0];
			let inden = data[ev]["list"][li]["indent"];
			
			// Checks if cType is in codeTypes array and if it is pushs it to extractedText array
			if (codeTypes.indexOf(cType) > -1){
				
				if (cType == 108){
					if (Param.includes("text_indicator") == true){
						tmpArr.push([{"codetype":cType, "indent":inden, "parameters":Param}]);
						continue;
					}
					continue;
				}
				
				if (cType == 105 || cType == 101){
					relatedChunk += 1;
				}
				if (cType == 0 && arrInArr.length > 0){
					relatedChunk = 0;
					tmpArr.push(arrInArr)
					arrInArr = [];
					continue;
				}
				if (cType == 101 && relatedChunk > 1 || cType == 105 && relatedChunk > 1){
					relatedChunk = 1;
					tmpArr.push(arrInArr);
					arrInArr = [];
					continue;
				}
				
				if (cType == 102 && relatedChunk > 1 || cType == 102 && relatedChunk == 0){
					tmpArr.push([{"codetype":cType, "indent":inden, "parameters":Param}])
					}
				if (relatedChunk == 1 && cType != 101 && cType != 105 ){
					arrInArr.push({"codetype":cType, "indent":inden, "parameters":Param});
					continue;
				}
			}
		}
		extractedText.push({"id":ev, "list":tmpArr});
	}
	return extractedText;
}

// short for actors armors classes skills states troops weapons
// can't think of a better name for this
function aacesstw_Parser(jData){
	
	let data = jData;
	let extractedText = [];
	
	let validKeys = ["name","description","message1","message2","message3","message4","nickname"];
	let fileKeys  = Object.keys(data[1]);
	
	let presentValidKeys = [];
	let vKeyLength = validKeys.length;
	
	
	//Finds valid keys in current file
	for (let v = 0; v < vKeyLength; ++v){
		if (fileKeys.indexOf(validKeys[v]) > -1){
			presentValidKeys.push(validKeys[v])
		}
	}
	
	let pVKLength = presentValidKeys.length
	let idLength = data.length
	for (let id = 0; id < idLength; ++id){
		if (data[id] === null){continue;}
		let tmpArr = [];
		for (let k = 0; k < pVKLength; ++k){
			tmpArr.push([presentValidKeys[k], data[id][presentValidKeys[k]]]);
		}
		const obj = Object.fromEntries(tmpArr);
		extractedText.push({"id":id, "list":[obj]});
	}
	return extractedText;
}

function system_Parser(jData){
	
	let data = jData;
	let extractedText = [];
	let tmpArr = [];
	
	tmpArr.push([
		Object.values(data['gameTitle']),
		Object.values(data['skillTypes']),
		Object.values(data['weaponTypes']),
		Object.values(data['equipTypes']),
		Object.values(data['armorTypes']),
		Object.values(data['elements']),
		Object.values(data['terms']['basic']),
		Object.values(data['terms']['commands']),
		Object.values(data['terms']['params']),
		Object.values(data['terms']['messages'])
		]);
	
	extractedText.push({"id":0, "list":tmpArr});
	return extractedText;
}

module.exports = {system_Parser, aacesstw_Parser, ce_Parser, map_Parser};

File: Py/parsers/Parsers.py

def loop_id (data_id_array, type):
	parsed_data = []

	for i in data_id_array:

		if i == None:
			parsed_data.append("")
			continue

		if type == 0:
			parsed_list = parse_list(i["list"])
			tmp_dict = {"id":i["id"], "indexs":parsed_list[0], "list":parsed_list[1]}
			parsed_data.append(tmp_dict)
			continue

		if type == 1:
			parsed_pages = loop_page(i["pages"])
			tmp_dict = {"id":i["id"], "indexs":parsed_pages[0], "list":parsed_pages[1]}
			parsed_data.append(tmp_dict)

		else:
			parsed_obj = parse_other(i)
			tmp_dict = {"id":i["id"], "list":parsed_obj}
			parsed_data.append(tmp_dict)

	return parsed_data

def loop_page (data_id_page_array):
	
	parsed_lists = []

	indexs = []

	for i in data_id_page_array:

		list_parse = parse_list(i["list"])
		indexs.append(list_parse[0])
		parsed_lists.append(list_parse[1])

	return [indexs, parsed_lists]

def parse_list (data_list_array):

	parsed_data    = []
	indexs         = []

	related_data   = []
	related_indexs = []

	is_related   = False

	for i in enumerate(data_list_array):

		ctype = i[1]["code"]
		entry = repr(i[1])

		match ctype:

			case 101 | 105 | 0:
				if is_related == False:
					is_related = True
					related_indexs.append(i[0])

				else:
					if ctype == 0:
						is_related = False
					parsed_data.append(related_data)
					related_indexs.append(i[0])
					indexs.append(related_indexs)
					related_indexs = []
					related_data = []

			case 401 | 405:
				related_data.append(entry)

			case 102:
				if is_related == True:
					related_data.append(entry)
				else:
					parsed_data.append([entry])
					indexs.append([i[0]])

			case 108:
				if "text_indicator" in entry["parameters"][0]:
					parsed_data.append([entry])
					indexs.append([i[0]])

			case _:
				if is_related == True:
					is_related = False
					parsed_data.append(related_data)
					related_indexs.append(i[0])
					indexs.append(related_indexs)
					related_indexs = []
					related_data = []

	return [indexs, parsed_data]

def parse_other (data_id_obj):
	
	VALID_KEYS = ["description","name","message1","message2","message3","message4","nickname"]

	parsed_data = {}

	for i in VALID_KEYS:
		if i in data_id_obj:
			parsed_data[i] = data_id_obj[i] 

	return parsed_data

def parse_system (data_sys):

	data = {}

	data["gameTitle"]         = data_sys["gameTitle"]
	data["skillTypes"]        = data_sys["skillTypes"]
	data["weaponTypes"]       = data_sys["weaponTypes"]
	data["equipTypes"]        = data_sys["equipTypes"]
	data["armorTypes"]        = data_sys["armorTypes"]
	data["elements"]          = data_sys["elements"]
	data["terms"]["basic"]    = data_sys["terms"]["basic"]
	data["terms"]["commands"] = data_sys["terms"]["commands"]
	data["terms"]["params"]   = data_sys["terms"]["params"]
	data["terms"]["messages"] = data_sys["terms"]["messages"]

	return data